### ISSUE TITLE
Folding sections in code blocks

### ISSUE THREAD

**Issue by brendanzab** [facebook/docusaurus#2215]
## 🚀 Feature

(A clear and concise description of what the feature is.)

## Have you read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Yes

## Motivation

Sometimes it's nice to be able to fold away irrelevant sections in a full code example, so that you can focus on the important parts.

## Pitch

For example, if I wrote a Rust function:

```rust
fn main() {
    // sample:start
    println!("hello!");
    // sample:end
}
```

I'd want it to display as (note the indentation):

```rust
println!("hello!");
```

There would be a button on the code block to allow this to be expanded to the full code:

```rust
fn main() {
    println!("hello!");
}
```

### Example: Kotlin's documentation

For a live example, you can check out the Kotlin docs. They seem to have a nice UI for unfolding examples. For example, [here is their documentation on defining variables](https://kotlinlang.org/docs/reference/basic-syntax.html#defining-variables):

- Folded:

  <img width="739" alt="Screen Shot 2020-01-15 at 1 52 20 pm" src="https://user-images.githubusercontent.com/695077/72400903-b50b5280-379e-11ea-99cb-a33f151bdee6.png">
- Unfolded:

  <img width="735" alt="Screen Shot 2020-01-15 at 1 52 28 pm" src="https://user-images.githubusercontent.com/695077/72400909-b9d00680-379e-11ea-88e9-ed87c1a56470.png">
- Markdown source:

  <img width="774" alt="Screen Shot 2020-01-15 at 1 47 32 pm" src="https://user-images.githubusercontent.com/695077/72400913-bdfc2400-379e-11ea-96b2-cc6603171c08.png">

**Comment by brendanzab** [comment]
I tried posting this [on Canny](https://docusaurus-2.netlify.com/feedback), but I ran into an error:


<details>
  <summary>Screenshot</summary>
<img width="306" alt="Screen Shot 2020-01-15 at 2 03 47 pm" src="https://user-images.githubusercontent.com/695077/72401415-0bc55c00-37a0-11ea-934b-14dd33e3f5c2.png">
</details>

**Comment by elviswolcott** [comment]
I have built something like this before in Docusaurus v1, I can take a tool at itif it's something the maintainers think is worth having as a core feature.

**Comment by lex111** [comment]
I do not think that this feature should be included in core of D2. After all, you can create a plugin and enable it on demand, right?

**Comment by yangshun** [comment]
It could be a fun thing to work on, but lower priority. The main reason why Kotlin docs needed it was because they allowed running code in the browser, and hence they needed to include a bunch of boilerplate runner code which was not relevant to the sample.

**Comment by brendanzab** [comment]
Yeah no worries!

It can also be helpful if you're running your docs through CI, and want to have the setup code there to ensure all examples are runnable.

**Comment by Josh-Cena** [comment]
This will be closed without getting implemented in core. See https://github.com/facebook/docusaurus/pull/5783#issuecomment-1117161022

In the near future, I will send a PR to dogfood this on our own website, which can be used as an inspiration. I'll keep this open for now until I make that demo.
